urid/src/cache.rs

```diff
@@ -0,0 +1,80 @@
+use crate::feature::*;
+use crate::URID;
+use core::UriBound;
+use fnv::FnvHashMap;
+use std::collections::{hash_map, HashMap};
+use std::ffi::CStr;
+
+pub struct CachedMap<'a> {
+    map_feature: Map<'a>,
+    cache: FnvHashMap<*const i8, URID>,
```

Comment:
I'm not sure what the use is to put an hashmap in front of the `Map` feature, which is probably another kind of hashmap anyway?

While there's not too much overhead since the key is a pointer and it's driven by `fnv`, I believe it's it quite a bit still (although that could be benchmarked), and the most important issue is that the cache may allocate in the realtime thread, while the derived-struct approach never does, because every single URI is pre-mapped at initialization time.

Reply:
The `CachedMap` is a compromise between the flexibility of the raw mapping feature and the speed of storing every URID in a struct.

First of all, we don't know what's behind the raw mapping feature. It might just be a map, but it might also be a very complex and inefficient algorithm or even a request to a web server! Therefore, plugin developers should always cache URIDs in a way that they know how fast it is.

The LV2 examples recommend/suggest to store URIDs in a special struct since this is the fastest way to access them. However, this requires quite some boilerplate code and manual intervention, which is not in the style of Rust. Therefore I was looking for the fastest, yet idiomatic way to access URIDs and this is what I came up with:

First of all, this structure maps types that implement `UriBound` to URIDs, not simply strings. This way, you can use the pointer to the URI of that type as the key, which is static and therefore always valid and the same. I've also switched to the [FNV hashing algorithm](https://en.wikipedia.org/wiki/Fowler%E2%80%93Noll%E2%80%93Vo_hash_function), because it has a runtime that is exactly in O(n), where n is the size of the key in bytes. Since the size of a pointer is always the same while the plugin is running, hashing the keys is actually in O(1), which is conform to the definition of [being hard rt capable](http://lv2plug.in/ns/lv2core/lv2core.html#hardRTCapable). Also, if you only use `try_urid_of` in the Audio Class (which you should), the cache will never allocate memory, since it does not have mutable access.

I know, using this cached map isn't as fast as storing URIDs in a struct, but it is convenient and fast enough to be used in a real-time-ish environment. However, we should also benchmark it in order to assure that my theories are correct! :smiley:

Reply:
The main issue I have with using something like `try_urid_of` is that it is a fallible operation, and I have no idea what you (as a plugin developer) can do if you cannot find your UUID. Either you panic (and bring the whole plugin instance down and put the host in an inconsistent state, because `run()` is not supposed to be fallible), or you try to handle the error somewhat and you'll probably just output silence, which I believe is even worse.

In order to guarantee no allocations and no runtime failures with this system, plugin developers would have to "prepare" the cache at initialization time by doing something like this: 

```rust
let _ = cached_map.urid_of::<Thing1>()?;
let _ = cached_map.urid_of::<Thing2>()?;
let _ = cached_map.urid_of::<Thing3>()?;
let _ = cached_map.urid_of::<Thing4>()?;
// [...]
```

Which brings back similar levels of boilerplate than we had before:

```rust
#[derive(URIDCache)]
struct MyCache {
  thing1: URIDOf<Thing1>,
  thing2: URIDOf<Thing2>,
  thing3: URIDOf<Thing3>,
  thing4: URIDOf<Thing4>
// [...]
}
```

… but with a few more downsides: on top of reduced perfomance, a forgotten declaration in the `CachedMap` would either allocate or straight-up crash at runtime, in probably very hardly debuggable situations, while the `URIDCache` struct will catch the issue at compile time.

Moreover, with the `URIDCache` struct we could also easily bundle several URID cache declarations together (such as all of the standard `Atom` types for example) to further reduce the boilerplate:

```rust
#[derive(URIDCache)]
struct MyCache {
  things: lv2::things::AllTheThings
}
```

Reply:
I had these concerns too, but didn't find a way to do it differently. However, you did! Using this `URIDOf<Something>` and the `URIDCache` trait looks very fast and comfortable. I will implement them soon!

Reply:
If you do, feel free to steal those implementations from [my old repository on GitLab](https://gitlab.com/prokopyl/lv2-rs/tree/master/lv2/src/urid) where I already implemented them (the derive macros can be found [here](https://gitlab.com/prokopyl/lv2-rs/tree/master/lv2/lv2-derive/src) as well).

I can also implement those rather easily if you'd like, since I already implemented them I know what little adjustments are needed to integrate them here! :wink:

Reply:
Suggestion implemented by commit 3bb9276ba2bbe3025983ad2f33dade6db39ea646.

Reply:
Looks neat :+1: